# 1. Sum

The sum of an array is the total obtained by adding all its elements together.

NumPy provides `np.sum()` to calculate the total efficiently.

### Example

If the sales are:

[100, 200, 300]

Total sales = 100 + 200 + 300 = 600

### Business Example

A company can use the sum to calculate total sales or total revenue.

### AI/ML Use Case

Sum is commonly used when calculating totals, aggregating features, and performing mathematical operations during data preprocessing.

In [1]:
import numpy as np

sales = np.array([100, 200, 300])

total = np.sum(sales)

print("Total:", total)

Total: 600


### Explanation

`np.sum()` adds all values in the array.

Expected output:

Total: 600

# 2. Mean

Mean represents the average value of a dataset.

It is calculated by adding all values and dividing the result by the number of values.

NumPy provides `np.mean()` for this calculation.

### Example

For:

[10, 20, 30]

Mean = (10 + 20 + 30) / 3 = 20

### Business Example

A business can calculate the average monthly sales.

### AI/ML Use Case

Mean is used in data analysis, feature preprocessing, and normalization.

In [2]:
import numpy as np

data = np.array([10, 20, 30])

mean = np.mean(data)

print("Mean:", mean)

Mean: 20.0


### Explanation

`np.mean()` calculates the average of the array.

Expected output:

Mean: 20.0

# 3. Median

The median is the middle value after arranging the data in order.

If there is an even number of values, the median is the average of the two middle values.

NumPy provides `np.median()`.

### Example

For:

[10, 20, 30, 40, 50]

The middle value is 30.

Therefore, the median is 30.

### Business Example

Median salary can give a better idea of a typical employee salary when a few employees have extremely high salaries.

### AI/ML Use Case

Median is useful when analyzing data containing outliers.

In [3]:
import numpy as np

data = np.array([10, 20, 30, 40, 50])

median = np.median(data)

print("Median:", median)

Median: 30.0


### Explanation

`np.median()` finds the middle value of the dataset.

Expected output:

Median: 30.0

# 4. Variance

Variance measures how spread out the values are from their mean.

A larger variance means the values are more spread out.

NumPy provides `np.var()` to calculate variance.

### Example

Consider:

[10, 20, 30]

The mean is 20.

The values are 10 units below, 0 units away, and 10 units above the mean.

Variance summarizes these squared differences.

### Business Example

A company can use variance to understand how much daily sales fluctuate.

### AI/ML Use Case

Variance is important in feature analysis and statistical modeling.

In [4]:
import numpy as np

data = np.array([10, 20, 30])

variance = np.var(data)

print("Variance:", variance)

Variance: 66.66666666666667


### Explanation

`np.var()` calculates the variance of the values.

Expected output:

Variance: 66.66666666666667

A higher value indicates greater spread in the data.

# 5. Standard Deviation

Standard deviation measures how much values typically differ from the mean.

It is the square root of variance.

NumPy provides `np.std()` to calculate it.

### Example

If two datasets have the same mean but one has a larger standard deviation, that dataset has more spread.

### Business Example

Standard deviation can show how much daily sales normally vary from average sales.

### AI/ML Use Case

It is commonly used for feature scaling and understanding data distribution.

In [5]:
import numpy as np

data = np.array([10, 20, 30])

std = np.std(data)

print("Standard Deviation:", std)

Standard Deviation: 8.16496580927726


### Explanation

`np.std()` calculates the standard deviation.

Expected output will be approximately:

Standard Deviation: 8.16

# 6. Minimum

The minimum is the smallest value present in an array.

NumPy provides `np.min()` to find it.

### Example

For:

[25, 10, 40, 15]

The minimum value is 10.

### Business Example

A company could use minimum sales to find its lowest sales day.

### AI/ML Use Case

Minimum values can be used during data exploration and feature preprocessing.

In [6]:
import numpy as np

data = np.array([25, 10, 40, 15])

minimum = np.min(data)

print("Minimum:", minimum)

Minimum: 10


### Explanation

`np.min()` returns the smallest element.

Expected output:

Minimum: 10

# 7. Maximum

The maximum is the largest value present in an array.

NumPy provides `np.max()` to find it.

### Example

For:

[25, 10, 40, 15]

The maximum value is 40.

### Business Example

A company could identify its highest sales day using the maximum.

### AI/ML Use Case

Maximum values are useful during data analysis and feature preprocessing.

In [7]:
import numpy as np

data = np.array([25, 10, 40, 15])

maximum = np.max(data)

print("Maximum:", maximum)

Maximum: 40


### Explanation

`np.max()` returns the largest element.

Expected output:

Maximum: 40

# 8. Argmin

`argmin` returns the index of the smallest value in an array.

It tells us where the minimum value is located rather than returning the value itself.

### Example

For:

[25, 10, 40, 15]

The minimum value is 10.

Its index is 1.

### Business Example

A company could find which position in a list represents the lowest sales.

### AI/ML Use Case

Argmin can be used to find the position of the smallest distance, error, or cost.

In [8]:
import numpy as np

data = np.array([25, 10, 40, 15])

index = np.argmin(data)

print("Index of minimum:", index)
print("Minimum value:", data[index])

Index of minimum: 1
Minimum value: 10


### Explanation

`np.argmin()` returns the index of the smallest value.

Expected output:

Index of minimum: 1
Minimum value: 10

# 9. Argmax

`argmax` returns the index of the largest value in an array.

It tells us the position of the maximum value.

### Example

For:

[25, 10, 40, 15]

The maximum value is 40.

Its index is 2.

### AI/ML Use Case

Argmax is commonly used to select the class with the highest predicted probability in classification models.

In [9]:
import numpy as np

data = np.array([25, 10, 40, 15])

index = np.argmax(data)

print("Index of maximum:", index)
print("Maximum value:", data[index])

Index of maximum: 2
Maximum value: 40


### Explanation

`np.argmax()` returns the index of the largest value.

Expected output:

Index of maximum: 2
Maximum value: 40

# 10. Percentile

A percentile tells us the value below which a certain percentage of observations fall.

For example, the 50th percentile represents the median.

NumPy provides `np.percentile()`.

### Example

For:

[10, 20, 30, 40, 50]

The 50th percentile is 30.

### Business Example

A company can use percentiles to understand employee salaries or customer spending.

### AI/ML Use Case

Percentiles are useful for understanding distributions and detecting extreme values.

In [10]:
import numpy as np

data = np.array([10, 20, 30, 40, 50])

p50 = np.percentile(data, 50)

print("50th Percentile:", p50)

50th Percentile: 30.0


### Explanation

`np.percentile(data, 50)` calculates the 50th percentile.

Expected output:

50th Percentile: 30.0

# 11. Dot Product

The dot product combines corresponding elements from two arrays by multiplying them and then adding the results.

For vectors A and B:

A · B = a1b1 + a2b2 + a3b3

### Example

A = [1, 2, 3]

B = [4, 5, 6]

Dot product:

(1×4) + (2×5) + (3×6) = 32

### AI/ML Use Case

Dot products are fundamental in machine learning models, including linear regression, neural networks, and similarity calculations.

In [11]:
import numpy as np

A = np.array([1, 2, 3])
B = np.array([4, 5, 6])

result = np.dot(A, B)

print("Dot Product:", result)

Dot Product: 32


### Explanation

`np.dot()` multiplies corresponding values and adds the results.

Expected output:

Dot Product: 32

# 12. Matrix Multiplication

Matrix multiplication combines two matrices by multiplying rows of the first matrix with columns of the second matrix.

The number of columns in the first matrix must match the number of rows in the second matrix.

NumPy provides `np.matmul()` and the `@` operator.

### Example

A =

[1 2]

[3 4]

B =

[5 6]

[7 8]

The result is:

[19 22]

[43 50]

### AI/ML Use Case

Matrix multiplication is heavily used in neural networks and other machine learning algorithms.

In [12]:
import numpy as np

A = np.array([
    [1, 2],
    [3, 4]
])

B = np.array([
    [5, 6],
    [7, 8]
])

result = A @ B

print("Matrix Multiplication:")
print(result)

Matrix Multiplication:
[[19 22]
 [43 50]]


### Explanation

The `@` operator performs matrix multiplication.

Expected output:

[[19 22]

 [43 50]]

Matrix multiplication is important for transformations and calculations inside machine learning models.

# 13. Transpose

The transpose of a matrix switches its rows and columns.

Rows become columns, and columns become rows.

NumPy provides `.T` to calculate the transpose.

### Example

Original matrix:

[1 2 3]

[4 5 6]

Transpose:

[1 4]

[2 5]

[3 6]

### AI/ML Use Case

Transpose is frequently used in matrix calculations and machine learning algorithms.

In [13]:
import numpy as np

matrix = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

transpose = matrix.T

print("Original:")
print(matrix)

print("Transpose:")
print(transpose)

Original:
[[1 2 3]
 [4 5 6]]
Transpose:
[[1 4]
 [2 5]
 [3 6]]


### Explanation

`.T` switches the rows and columns of the matrix.

A 2 × 3 matrix becomes a 3 × 2 matrix.

### Explanation

`.T` switches the rows and columns of the matrix.

A 2 × 3 matrix becomes a 3 × 2 matrix.

# 14. Basic Linear Algebra Operations

NumPy provides tools for common linear algebra operations such as:

- Matrix multiplication
- Dot products
- Transpose
- Matrix inverse
- Determinant
- Eigenvalues and eigenvectors

These operations are useful when working with vectors and matrices.

### Example

Consider the matrix:

A =

[1 2]
[3 4]

We can calculate its determinant and inverse using NumPy.

### AI/ML Use Case

Linear algebra is a foundation of machine learning. Matrices and vectors are used in regression, neural networks, PCA, and many other algorithms.

In [14]:
import numpy as np

A = np.array([
    [1, 2],
    [3, 4]
])

determinant = np.linalg.det(A)
inverse = np.linalg.inv(A)

print("Matrix:")
print(A)

print("Determinant:", determinant)

print("Inverse:")
print(inverse)

Matrix:
[[1 2]
 [3 4]]
Determinant: -2.0000000000000004
Inverse:
[[-2.   1. ]
 [ 1.5 -0.5]]


### Explanation

`np.linalg.det()` calculates the determinant of the matrix.

`np.linalg.inv()` calculates the inverse of the matrix.

The determinant will be approximately:

-2.0

The inverse will be approximately:

[[-2.   1. ]

 [ 1.5 -0.5]]

These linear algebra operations are widely used in AI/ML algorithms involving matrices and transformations.